test/unit/matrix/test_matrix_output.cpp

```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/matrix/print_numpy.h"
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+#include <gtest/gtest.h>
+#include <hpx/include/util.hpp>
+#include <hpx/local/future.hpp>
+
+#include "dlaf_test/comm_grids/grids_6_ranks.h"
+#include "dlaf_test/matrix/util_matrix.h"
+#include "dlaf_test/util_types.h"
+
+using namespace dlaf;
+using namespace dlaf::matrix;
+
+using dlaf::matrix::test::set;
+
+::testing::Environment* const comm_grids_env =
+    ::testing::AddGlobalTestEnvironment(new dlaf::test::CommunicatorGrid6RanksEnvironment);
+
+template <typename Type>
+class MatrixOutputLocalTest : public ::testing::Test {};
+
+TYPED_TEST_SUITE(MatrixOutputLocalTest, dlaf::test::MatrixElementTypes);
+
+struct TestSizes {
+  LocalElementSize size;
+  TileElementSize block_size;
+};
+
+const std::vector<TestSizes> sizes({
+    {{0, 0}, {2, 2}},
+    {{6, 6}, {2, 2}},
+    {{6, 6}, {3, 3}},
+    {{8, 8}, {3, 3}},
+    {{9, 7}, {3, 3}},
+});
+
+template <class T>
+T pattern_values(const GlobalElementIndex& index) {
+  SizeType i = index.row();
+  SizeType j = index.col();
+  return dlaf::test::TypeUtilities<T>::element(i + j, j - i);
+}
+
+template <class T>
+struct numpy_type {
+  static constexpr auto name = "single";
+
+  static T deserialize(const std::string& value_str) {
+    std::istringstream deserializer(value_str);
+    T value;
+    deserializer >> value;
+    return value;
+  }
+};
+
+template <class T>
+struct numpy_type<std::complex<T>> {
+  static constexpr auto name = "csingle";
+
+  static std::complex<T> deserialize(const std::string& value_str) {
+    const std::string regex_sign{"([\\+\\-])"};              // <capture sign>
+    const std::string regex_real{"([0-9]+(?:\\.[0-9]+)?)"};  // <floating-point value> without sign
+    const std::regex regex_complex{"\\s*" + regex_sign + "?\\s*" + regex_real +  // <(+/-)> <real>
+                                   "\\s*" + regex_sign + "\\s*" + regex_real +   // <+/-> <imag>
+                                   "[jJ]"};                                      // j (or J)
+
+    std::smatch matches;
+    DLAF_ASSERT(std::regex_match(value_str, matches, regex_complex), value_str,
+                "does not look like a complex number");
+
+    const auto real = numpy_type<T>::deserialize(matches[1].str() + matches[2].str());
+    const auto imag = numpy_type<T>::deserialize(matches[3].str() + matches[4].str());
+
+    return {real, imag};
+  }
+};
+
+template <class T>
+::testing::AssertionResult parseAndCheckTile(const std::string numpy_output,
+                                             const Tile<const T, Device::CPU>& tile) {
+  const std::regex regex_syntax{
+      "np\\.array\\(\\[(.*)\\],\\s*dtype=np\\.(\\w+)\\)"  // np.array([<values>], dtype=np.<type>)
+      "\\.reshape\\((\\d+),\\s*(\\d+)\\)\\.T\\s*"};       // .reshape(<n>, <n>).T
```

Comment:
The unit tests are quite comprehensive but I find that their complexity exceeds that of what they are testing. I think it's worth aiming for simplicity. I had to google to understand all the parts of the regex here, it's very likely that I will have to do it again in a few weeks if I come back to this part of the code.

Instead of using `std::regex`, why don't you instead write down the complete expected string explicitly and compare it to what you get? I know it's tempting to automate it but I find it makes more sense to test several "interesting" cases, instead of a large number of similar cases. Sometimes it does make sense to automate and test in bulk (like for Cholesky or the triangular solver), but I think we can achieve a reasonable confidence in correctness by being more explicit about the expected results and by picking a few "interesting" cases here.

However, I can understand if you would like to keep the code as it is, but I do think there is certain maintainability and comprehensibility costs that have to be taken into account.

Reply:
Yes, I agree, it's way more "complex" the testing code than the code to be tested.

For me it's ok, we can pick up some specific local/distributed cases and just test them.

If others agree I can simplify them, it's not a problem. @rasolca @Sely85

Reply:
I agree that the `regex `expressions are really "complex" to understand. A "simplified" version of unit tests (when it is possible) like that used in `test_matrix_local.cpp` looks more clear to me.